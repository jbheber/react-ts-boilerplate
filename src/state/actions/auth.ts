import { createAction } from 'redux-act'

export const LOGOUT_INIT = createAction<string>('LOGOUT_INIT')
export const LOGOUT_SUCCESS = createAction('LOGOUT_SUCCESS')
export const LOGOUT_FAIL = createAction('LOGOUT_FAIL')
