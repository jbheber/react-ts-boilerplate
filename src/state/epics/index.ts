import { combineEpics } from 'redux-observable'
import global from './global'

export default combineEpics(...[...global])
