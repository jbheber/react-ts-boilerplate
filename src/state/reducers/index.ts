import { combineReducers } from 'redux'
import { persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'

import { globalReducer } from './global'

export default combineReducers({
  global: persistReducer(
    {
      key: 'global',
      storage,
      blacklist: ['isChangingLanguage']
    },
    globalReducer
  )
  // Add reducers here
})
