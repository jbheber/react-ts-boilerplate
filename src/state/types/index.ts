export * from './global'
export * from './state'
