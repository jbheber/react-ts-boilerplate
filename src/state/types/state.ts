import { IGlobalState } from './global'

export interface IState {
  global: IGlobalState
}
