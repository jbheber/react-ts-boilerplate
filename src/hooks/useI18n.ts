import * as languages from './locales'

import { IState, Language } from '../state/types'

import { useMemo } from 'react'
import { useSelector } from 'react-redux'

interface ILanguageFiles {
  [key: string]: {
    [key: string]: string
  }
}

const getTranslate = (translations: { [key: string]: string }) => (
  key: string,
  ...params: string[]
) => {
  let value = translations[key] || key
  if (params) {
    params.forEach((p, idx) => {
      value = value.replace(`{${idx}}`, p)
    })
  }

  return value
}

/**
 * Custom hook to get localized strings.
 * Will return a function that gets the localized string for a given key.
 */
export const useI18n = () => {
  const language = useSelector<IState, Language>(state => state.global.language)
  const languageFile: ILanguageFiles = languages
  return useMemo(() => getTranslate(languageFile[language]), [
    language,
    languageFile
  ])
}
