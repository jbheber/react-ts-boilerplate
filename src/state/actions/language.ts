import { Language } from '../types'
import { createAction } from 'redux-act'

export const CHANGE_LANGUAGE_INIT = createAction<Language>(
  'CHANGE_LANGUAGE_INIT'
)
export const CHANGE_LANGUAGE_SUCCESS = createAction('CHANGE_LANGUAGE_SUCCESS')
export const CHANGE_LANGUAGE_FAIL = createAction<any>('CHANGE_LANGUAGE_FAIL')
export const SET_LANGUAGE = createAction<Language>('SET_LANGUAGE')
