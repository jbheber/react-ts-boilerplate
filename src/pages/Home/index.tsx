import './Home.scss'

import { IState, Language } from '../../state/types'
import { useDispatch, useSelector } from 'react-redux'

import { CHANGE_LANGUAGE_INIT } from '../../state/actions'
import { Dispatch } from 'redux'
import React from 'react'
import logo from './logo.svg'
import { useI18n } from '../../hooks'

function onClickHandler(dispatch: Dispatch<any>, language: Language) {
  const nextLang = Language.EN === language ? Language.ES : Language.EN
  return () => dispatch(CHANGE_LANGUAGE_INIT(nextLang))
}

const Home = () => {
  const t = useI18n()
  const dispatch = useDispatch()
  const language = useSelector<IState, Language>(state => state.global.language)
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p onClick={onClickHandler(dispatch, language)}>{t('CLICK_HERE')}</p>
        <a
          className="App-link"
          href="https://reactjs.org"
          target="_blank"
          rel="noopener noreferrer"
        >
          {t('LEARN_REACT')}
        </a>
      </header>
    </div>
  )
}

export default Home
