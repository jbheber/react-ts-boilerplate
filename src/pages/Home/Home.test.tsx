import * as i18n from '../../hooks/useI18n'
import * as reactRedux from 'react-redux'

import { CHANGE_LANGUAGE_INIT } from '../../state/actions'
import Home from '.'
import { Language } from '../../state/types'
import React from 'react'

describe('Home page', () => {
  let dispatchMock: jest.Mock<any, any>

  beforeEach(() => {
    dispatchMock = jest.fn()
    jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => dispatchMock)
    jest
      .spyOn(i18n, 'useI18n')
      .mockImplementation(() => (key, ...params) =>
        `${key.toUpperCase()} ${params.join('_')}`.trim()
      )
  })
  it('renders without crashing', () => {
    const { component } = mountWithProvider(<Home />)({
      global: { language: Language.EN }
    })
    expect(component).toMatchSnapshot()
  })

  it('Fires a change language on click `en` -> `es`', () => {
    const { component } = mountWithProvider(<Home />)({
      global: { language: Language.EN }
    })
    component.find('p').simulate('click')
    expect(dispatchMock).toBeCalledWith(CHANGE_LANGUAGE_INIT(Language.ES))
  })

  it('Fires a change language on click `es` -> `en`', () => {
    const { component } = mountWithProvider(<Home />)({
      global: { language: Language.ES }
    })
    component.find('p').simulate('click')
    expect(dispatchMock).toBeCalledWith(CHANGE_LANGUAGE_INIT(Language.EN))
  })
})
