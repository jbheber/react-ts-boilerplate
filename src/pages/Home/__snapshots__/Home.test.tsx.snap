// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page renders without crashing 1`] = `
<TestComponent
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Provider
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <Home>
          <div
            className="App"
          >
            <header
              className="App-header"
            >
              <img
                alt="logo"
                className="App-logo"
                src="logo.svg"
              />
              <p
                onClick={[Function]}
              >
                CLICK_HERE
              </p>
              <a
                className="App-link"
                href="https://reactjs.org"
                rel="noopener noreferrer"
                target="_blank"
              >
                LEARN_REACT
              </a>
            </header>
          </div>
        </Home>
      </Provider>
    </Router>
  </BrowserRouter>
</TestComponent>
`;
